vcd/metadata_openapi.go

```diff
@@ -0,0 +1,329 @@
+					Description:  fmt.Sprintf("Type of this metadata entry. One of: '%s', '%s', '%s'", types.OpenApiMetadataStringEntry, types.OpenApiMetadataNumberEntry, types.OpenApiMetadataBooleanEntry),
+					ValidateFunc: validation.StringInSlice([]string{types.OpenApiMetadataStringEntry, types.OpenApiMetadataNumberEntry, types.OpenApiMetadataBooleanEntry}, false),
+				},
+				"readonly": {
+					Type:        schema.TypeBool,
+					Optional:    true,
+					Default:     false,
+					Description: "True if the metadata entry is read only",
+				},
+				"domain": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					Default:      "TENANT",
+					Description:  "Only meaningful for providers. Allows them to share entries with their tenants. Currently, accepted values are: `TENANT`, `PROVIDER`",
+					ValidateFunc: validation.StringInSlice([]string{"TENANT", "PROVIDER"}, false),
+				},
+				"namespace": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Namespace of the metadata entry",
+				},
+				"persistent": {
+					Type:        schema.TypeBool,
+					Optional:    true,
+					Description: "Persistent metadata entries can be copied over on some entity operation",
+				},
+			},
+		},
+	}
+}
+
+// openApiMetadataCompatible allows to consider all structs that implement OpenAPI metadata handling to be the same type
+type openApiMetadataCompatible interface {
+	GetMetadata() ([]*types.OpenApiMetadataEntry, error)
+	GetMetadataByKey(namespace, key string) (*types.OpenApiMetadataEntry, error)
+	AddMetadata(metadataEntry types.OpenApiMetadataEntry) (*types.OpenApiMetadataEntry, error)
+	UpdateMetadata(namespace, key string, value interface{}) (*types.OpenApiMetadataEntry, error)
+	DeleteMetadata(namespace, key string) error
+}
+
+// createOrUpdateOpenApiMetadataEntryInVcd creates or updates OpenAPI metadata entries in VCD for the given resource, only if the attribute
+// metadata_entry has been set or updated in the state.
+func createOrUpdateOpenApiMetadataEntryInVcd(d *schema.ResourceData, resource openApiMetadataCompatible) error {
+	if !d.HasChange("metadata_entry") {
+		return nil
+	}
+
+	oldRaw, newRaw := d.GetChange("metadata_entry")
+	metadataToAdd, metadataToUpdate, metadataToDelete, err := getOpenApiMetadataOperations(oldRaw.(*schema.Set).List(), newRaw.(*schema.Set).List())
+	if err != nil {
+		return fmt.Errorf("could not calculate the needed metadata operations: %s", err)
+	}
+
+	// getOpenApiMetadataOperations retrieves keys and namespaces merged with a separator, this function
+	// splits the values in two: namespace and key, separately.
+	getKeyAndNamespace := func(namespacedKey string) (string, string, error) {
+		r := strings.Split(namespacedKey, "%%%") // Separator used by getOpenApiMetadataEntryMap
+		if len(r) == 2 {
+			return r[0], r[1], nil
+		}
+		return "", "", fmt.Errorf("bad formatting of metadata map key %s, this is a provider error", namespacedKey)
```

Comment:
Would be good to clear out this part `this is a provider error` - it is ambiguous. Does it want to say that Terraform provider VCD error?

Reply:
Yes. Ideally it should never happen, but if it does, I wanted to make it clear that it's an issue in our side, not VCD.

Reply:
Changed to a more concise error.

Reply:
As per the comment above, I've removed this function, not needed at all as this was overcomplicated.